state/StateFilter.F90

```diff
@@ -0,0 +1,151 @@
+#include "MAPL_ErrLog.h"
+module MAPL_StateFilterMod
+   use ESMF
+   use MAPL_ExceptionHandling
+   use MAPL_FieldUtils
+   use MAPL_StateArithmeticParserMod
+   use MAPL_StateMaskMod
+   use, intrinsic :: iso_fortran_env, only: REAL32, REAL64
+   implicit none
+   private
+
+   public MAPL_StateFilterItem
+   character(len=1), parameter :: var_placeholder = "@"
+
+   interface MAPL_StateFilterItem
+      module procedure MAPL_StateFilter_R4_2D
+      module procedure MAPL_StateFilter_R4_3D
+   end interface
+
+   contains
+
+   subroutine MAPL_StateFilter_R4_2D(state, config, itemName, array, rc)
+      type(ESMF_State), intent(inout) :: state
+      type(ESMF_Config), intent(inout) :: config
+      character(len=*), intent(in) :: itemName
+      real(REAL32), allocatable, intent(out) :: array(:,:)
+      integer, optional, intent(out) :: rc
+
+      integer :: status, rank
+      character(len=ESMF_MAXSTR) :: filter_expression, field_name
+      character(len=:), allocatable :: processed_expression
+      type(ESMF_Field) :: new_field, old_field
+      logical :: name_Present, default_Present
+      real(REAL32), pointer :: ptr2d_new(:,:), ptr2d_old(:,:)
+      type(ESMF_TYPEKIND_FLAG) :: tk
+      type(StateMask) :: mask
+
+      call ESMF_StateGet(state, itemName, old_field, _RC)
+      call ESMF_FieldGet(old_field, typeKind=tk, rank=rank, _RC) 
+      _ASSERT(tk==ESMF_TYPEKIND_R4,"wrong typekind when call MAPL_StateFilter")
+      _ASSERT(rank==2,"wrong rank when call MAPL_StateFilter")
+
+      call ESMF_FieldGet(old_field, 0, farrayPtr=ptr2d_old, _RC)
+      allocate(array( lbound(ptr2d_old,1):ubound(ptr2d_old,1) , lbound(ptr2d_old,2):ubound(ptr2d_old,2) ),  _STAT) 
+      array = ptr2d_old
+ 
+      call ESMF_ConfigFindLabel(config, "FILTER%"//trim(itemName)//":", isPresent=name_Present, _RC)
+      if (name_Present) then
+         call ESMF_ConfigGetAttribute(config, filter_expression, label="FILTER%"//trim(itemName)//":", _RC)
+      else
+         call ESMF_ConfigFindLabel(config, "FILTER%*:", isPresent=default_Present, _RC)
+         if (default_present) then
+            call ESMF_ConfigGetAttribute(config, filter_expression, label="FILTER%*:", _RC)
+         else
+            _RETURN(_SUCCESS)
+         end if
```

Comment:
```suggestion
         _RETURN_UNLESS(default_present)
         call ESMF_ConfigGetAttribute(config, filter_expression, label="FILTER%*:", _RC)
```

Reply:
I realize some of this is probably just pasted code from elsewhere.   I'll try to refrain from doing too much of this.  But ... it really jumped out at me.